temporal-testing/src/main/java/io/temporal/testing/internal/SDKTestWorkflowRule.java

```diff
@@ -299,6 +307,31 @@ public static void assertNoHistoryEvent(History history, EventType eventType) {
     }
   }
 
+  /** Waits till the end of the workflow task if there is a workflow task in progress */
+  public void waitForTheEndOfWFT(WorkflowExecution execution) {
+    WorkflowExecutionHistory initialHistory = getExecutionHistory(execution);
+
+    HistoryEvent lastEvent = initialHistory.getLastEvent();
+    if (isWFTInProgress(lastEvent)) {
```

Comment:
Could this loop possibly skip over a whole WFT and think it's still "in progress" because now we're on to the next one?

Would it be reasonable to use some kind of wake-up rather than the busy sleep?

Reply:
> Could this loop possibly skip over a whole WFT and think it's still "in progress" because now we're on to the next one?        
```
        for (previous = historyEventListIterator.previous();
            historyEventListIterator.hasPrevious() && previous.getEventId() > startEventId;
            previous = historyEventListIterator.previous()) {
          if (!isWFTInProgress(historyEventListIterator.previous())) {
            return;
          }
        }
```
No, it's not. We remember where we start and we check the last position up to the started position trying to sign something other than SCHEDULED and STARTED. If we find - we return. So, if the new task will be added and is in STARTED position for example - we will catch it.

> Would it be reasonable to use some kind of wake-up rather than the busy sleep?
It 100% will be if it's an SDK code. For test purposes busy waiting is fine I think.